core/src/main/scala/cats/instances/eq.scala

```diff
@@ -43,4 +45,25 @@ trait EqInstances extends kernel.instances.EqInstances {
       def product[A, B](fa: Eq[A], fb: Eq[B]): Eq[(A, B)] =
         (left, right) => fa.eqv(left._1, right._1) && fb.eqv(left._2, right._2)
     }
+
+  implicit val catsDeferForEq: Defer[Eq] =
+    new Defer[Eq] {
+      case class Deferred[A](fa: () => Eq[A]) extends Eq[A] {
+        override def eqv(x: A, y: A): Boolean = {
+          @tailrec
+          def loop(f: () => Eq[A]): Boolean =
```

Comment:
I really think there should be a:
```scala
lazy val evaluated: Eq[A] = loop(fa)
def loop(f: () => Eq[A]): Eq[A] = ...
```
so we don't need to run the loop on every comparison.

I see that you are doing the `lazy val` trick below, but what if the Deferred evaluates to another deferred? That is not an uncommon case.

Reply:
Makes sense. As I'm going to be in there already, do you want me to retrofit the other instances (like `catsSddDeferForFunction0`)?

Reply:
Now the bin-compat issues have been resolved, I think this is the last open question.

Reply:
why not follow up? This looks ready to merge. Let's bias to action and merge this and if you have time send a new PR

Reply:
Works for me. I can't make the merge happen anyway, so I figured if it was important, you'd circle back, and if not, it's all good anyway :)

I mostly dropped a ping because I'd run out of other stuff to do on the ticket, and it's a slow day otherwise 😆